plugins/woocommerce/includes/gateways/paypal/includes/class-wc-gateway-paypal-notices.php

```diff
@@ -76,35 +137,156 @@ public function add_paypal_migration_notice() {
 	 * @return bool
 	 */
 	protected static function paypal_migration_notice_dismissed() {
-		return get_user_meta( get_current_user_id(), 'dismissed_paypal_migration_completed_notice', true );
+		return self::is_notice_dismissed( self::PAYPAL_MIGRATION_NOTICE );
+	}
+
+	/**
+	 * Check if there's a flag indicating PayPal account restriction.
+	 *
+	 * @return bool
+	 */
+	private function has_account_restriction_flag(): bool {
+		return 'yes' === $this->gateway->get_option( 'paypal_account_restricted', 'no' );
+	}
+
+	/**
+	 * Set the flag indicating PayPal account restriction.
+	 *
+	 * @return void
+	 */
+	public static function set_account_restriction_flag(): void {
+		$gateway = WC_Gateway_Paypal::get_instance();
+		if ( $gateway && 'no' === $gateway->get_option( 'paypal_account_restricted', 'no' ) ) {
+			$gateway->update_option( 'paypal_account_restricted', 'yes' );
+		}
+	}
+
+	/**
+	 * Clear the flag indicating PayPal account restriction.
+	 *
+	 * @return void
+	 */
+	public static function clear_account_restriction_flag(): void {
+		$gateway = WC_Gateway_Paypal::get_instance();
+		if ( $gateway && 'yes' === $gateway->get_option( 'paypal_account_restricted', 'no' ) ) {
+			$gateway->update_option( 'paypal_account_restricted', 'no' );
+		}
+	}
+
+	/**
+	 * Handle PayPal order response to manage account restriction notices.
+	 *
+	 * This method is called via the 'woocommerce_paypal_standard_order_created_response' hook
+	 * and manages the account restriction flag based on PayPal API responses.
+	 *
+	 * Extensions can disable this feature using the filter:
+	 * add_filter( 'woocommerce_paypal_account_restriction_notices_enabled', '__return_false' );
+	 *
+	 * @param int      $http_code     The HTTP status code from the PayPal API response.
+	 * @param array    $response_data The decoded response data from the PayPal API, or null if decoding failed.
+	 * @param WC_Order $order         The WooCommerce order object.
+	 * @return void
+	 */
+	public function manage_account_restriction_status( int $http_code, array $response_data, WC_Order $order ): void {
+		wc_get_logger()->debug( 'PayPal account restriction status: ' . $http_code . ' for order: ' . $order->get_id() );
+		/**
+		 * Filters whether account restriction notices should be enabled.
+		 *
+		 * This filter allows extensions to opt out of the account restriction notice functionality.
+		 *
+		 * @since 10.4.0
+		 *
+		 * @param bool $enabled Whether account restriction notices are enabled. Default true.
+		 */
+		if ( ! apply_filters( 'woocommerce_paypal_account_restriction_notices_enabled', true ) ) {
+			return;
+		}
+
+		// Clear the restriction flag on successful responses.
+		if ( in_array( $http_code, array( 200, 201 ), true ) ) {
+			self::clear_account_restriction_flag();
+			return;
+		}
+
+		// Set the restriction flag for account-related errors.
+		if ( 422 === $http_code && is_array( $response_data ) ) {
```

Comment:
This block of code is a bit too "crowded". Maybe we could use the coalesce operator and create a variable or constant with the two issue constants.

Reply:
Done [d12fabd](https://github.com/woocommerce/woocommerce/pull/62058/commits/d12fabd9025f205c9e254bbab3da7fc0689f130b)